### PR TITLE
Token registry

### PR DESCRIPTION
**Pull request by poiresel**
## Description

Adding ability for AirSwap Swap to handle different token asset interfaces per #145 
Quick description:

- [X ] New features

## Changes

Implementation details. These are mainly to help your fellow developers.

## Tests

Basic unit tests in place 
4 passing

Need to update Swap unit tests as well

### CODE REVIEW

File: helpers/wrapper/migrations/2_deploy_contracts.js [hunk 1]
```diff
@@ -7,7 +8,8 @@ module.exports = deployer => {
   deployer.deploy(Types)
   deployer.link(Types, Swap)
   deployer.deploy(WETH9)
-  deployer
-    .deploy(Swap)
+  deployer.deploy(TokenRegistry)
+    .then(() => TokenRegistry.deployed())
+    .then(() => deployer.deploy(Swap, TokenRegistry.address))
```
Comment (alicevhenshaw) [inline helpers/wrapper/migrations/2_deploy_contracts.js]: why does this need to be in a `.then`? Can't it be done as a new `deployer.deploy` line like the deployments on lines 8-11?

File: helpers/wrapper/test/Wrapper-unit.js [hunk 2]
```diff
@@ -74,7 +74,8 @@ contract('Wrapper Unit Tests', async accounts => {
   }
 
   async function setupMockSwap() {
-    let swapTemplate = await Swap.new()
+    let mockTokenRegistry = await MockContract.new()
```
Comment (alicevhenshaw) [inline helpers/wrapper/test/Wrapper-unit.js]: I think you can just pass in an `accounts[i]` as the mock token registry as it isn't actually being used?

File: helpers/wrapper/test/Wrapper.js [hunk 3]
```diff
@@ -41,7 +43,11 @@ contract('Wrapper', async ([aliceAddress, bobAddress, carolAddress]) => {
     // link types to swap
     await Swap.link(Types, (await Types.new()).address)
     // now deploy swap
-    swapContract = await Swap.new()
+
+    const erc20asset = await ERC20Asset.new()
```
Comment (alicevhenshaw) [inline helpers/wrapper/test/Wrapper.js]: Please can the `a` be capitalised for consistency?

File: protocols/token-registry/contracts/TokenRegistry.sol [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+  // Mapping of bytes4 to contract interface type
+  mapping (bytes4 => IAsset) private assetMapping;
+
+  constructor() public {
+    owner = msg.sender;
+  }
+
+  function addToRegistry(bytes4 _kind, IAsset _asset) external {
+    require(msg.sender == owner, "NOT OWNER");
```
Comment (dmosites) [inline protocols/token-registry/contracts/TokenRegistry.sol]: Suggest `SENDER_MUST_BE_OWNER`

File: protocols/token-registry/contracts/TokenRegistry.sol [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    assetMapping[_kind] = _asset;
+    emit AddToRegistry(_kind, address(_asset));
+  }
+
+  function removeFromRegistry(bytes4 _kind) external {
+    require(msg.sender == owner, "NOT OWNER");
```
Comment (dmosites) [inline protocols/token-registry/contracts/TokenRegistry.sol]: Suggest `SENDER_MUST_BE_OWNER`

File: protocols/token-registry/contracts/TokenRegistry.sol (hunk 4 above)
Comment (alicevhenshaw) [reply protocols/token-registry/contracts/TokenRegistry.sol]: Why do we not just use `Ownable`?

File: protocols/token-registry/test/TestTokenRegistry-unit.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
+const TokenRegistry = artifacts.require('TokenRegistry')
+const { takeSnapshot, revertToSnapShot } = require('@airswap/test-utils').time
+const { reverted, equal, emitted } = require('@airswap/test-utils').assert
+
+contract('TokenRegistry Unit Tests', async accounts => {
+  const owner = accounts[0]
+  const nonOwner = accounts[1]
+  const erc20asset = accounts[4]
```
Comment (alicevhenshaw) [inline protocols/token-registry/test/TestTokenRegistry-unit.js]: Why `[4]`?

File: protocols/token-registry/test/TestTokenRegistry-unit.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+  beforeEach(async () => {
+    let snapShot = await takeSnapshot()
+    snapshotId = snapShot['result']
+  })
+
+  afterEach(async () => {
+    await revertToSnapShot(snapshotId)
+  })
+
+  before('Deploy TokenRegistry Factory', async () => {
```
Comment (alicevhenshaw) [inline protocols/token-registry/test/TestTokenRegistry-unit.js]: Why does this say `Factory`?

File: protocols/token-registry/test/TestTokenRegistry-unit.js [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+          from: owner,
+        }),
+        'AddToRegistry'
+      )
+
+      await emitted(
+        await tokenregistry.removeFromRegistry('0x80ac58cd', { from: owner }),
+        'RemoveFromRegistry'
+      )
+    })
```
Comment (alicevhenshaw) [inline protocols/token-registry/test/TestTokenRegistry-unit.js]: Should test functionality of `getAsset` - both when the mapping has a result and when it doesn't

File: protocols/swap/test/Swap.js [hunk 9]
```diff
@@ -81,9 +89,22 @@ contract('Swap', async accounts => {
     it('Deployed test contract "DAI"', async () => {
       tokenDAI = await FungibleToken.new()
     })
+
+    it('Set up TokenRegistry', async () => {
```
Comment (alicevhenshaw) [inline protocols/swap/test/Swap.js]: Add tests for when swapping tokens with a `kind` that isn't in the token registry

File: helpers/delegate-frontend/test/DelegateFrontend.js [hunk 10]
```diff
@@ -54,7 +56,11 @@ contract('DelegateFrontend', async accounts => {
       // link types to swap
       await Swap.link(Types, (await Types.new()).address)
       // now deploy swap
-      swapContract = await Swap.new()
+
+      const erc20asset = await ERC20Asset.new()
+      const tokenRegistry = await TokenRegistry.new()
+      tokenRegistry.addToRegistry('0x277f8169', erc20asset.address)
```
Comment (ejwessel) [inline helpers/delegate-frontend/test/DelegateFrontend.js]: generate the selector.

**Comment by ejwessel** [comment]
FYI there has also been some discussion about naming 
`IAsset` → `ITransferHandler`
`TokenRegistry` → `TransferHandlerRegistry`
`addToRegistry` etc → `addHandler`
`removeFromRegistry` etc → `removeHandler`

File: protocols/swap/contracts/Swap.sol [hunk 11]
```diff
@@ -17,6 +17,8 @@
 pragma solidity 0.5.10;
 pragma experimental ABIEncoderV2;
 
+import "@airswap/token-registry/contracts/interfaces/IAsset.sol";
```
[the remark this answers is not in the document]
Comment (alicevhenshaw) [reply protocols/swap/contracts/Swap.sol]: It's used on line 300

File: helpers/wrapper/migrations/2_deploy_contracts.js (hunk 1 above)
Comment (poiresel) [reply helpers/wrapper/migrations/2_deploy_contracts.js]: A tokenRegistry address needs to be accessible similar to how swap is needed for wrapper. deployer.deploy returns a promise so just having a list of promises would not be sufficient, I chained them so they were getting each others results.

File: protocols/token-registry/contracts/TokenRegistry.sol (hunk 4 above)
Comment (poiresel) [reply protocols/token-registry/contracts/TokenRegistry.sol]: done

File: protocols/token-registry/contracts/TokenRegistry.sol (hunk 5 above)
Comment (poiresel) [reply protocols/token-registry/contracts/TokenRegistry.sol]: made it onlyOwner

File: helpers/wrapper/migrations/2_deploy_contracts.js (hunk 1 above)
Comment (alicevhenshaw) [reply helpers/wrapper/migrations/2_deploy_contracts.js]: Yeah makes sense. You can fix it using `await` on each line but isn't a problem as is

File: protocols/token-registry/test/TestTokenRegistry-unit.js (hunk 6 above)
Comment (poiresel) [reply protocols/token-registry/test/TestTokenRegistry-unit.js]: don't have a good reason, making it accounts[2]

File: protocols/token-registry/test/TestTokenRegistry-unit.js (hunk 7 above)
Comment (poiresel) [reply protocols/token-registry/test/TestTokenRegistry-unit.js]: removing

File: protocols/token-registry/contracts/TokenRegistry.sol (hunk 4 above)
Comment (alicevhenshaw) [reply protocols/token-registry/contracts/TokenRegistry.sol]: I don't see the change - have you pushed?

File: protocols/swap/test/Swap.js (hunk 9 above)
Comment (poiresel) [reply protocols/swap/test/Swap.js]: good catch!

File: protocols/token-registry/contracts/TokenRegistry.sol [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    delete assetMapping[_kind];
+    emit RemoveFromRegistry(_kind, address(assetMapping[_kind]));
+  }
+
+  function getAsset(bytes4 _kind) external view returns (IAsset) {
+    return assetMapping[_kind];
+  }
+}
```
[the remark this answers is not in the document]
Comment (poiresel) [reply protocols/token-registry/contracts/TokenRegistry.sol]: was waiting on thoughts on tokenregistry design, but ultimately yes.

File: protocols/swap/contracts/Swap.sol [hunk 13]
```diff
@@ -311,12 +297,8 @@ contract Swap is ISwap {
...
   ) internal {
-    if (_kind == ERC721_INTERFACE_ID) {
-      // Attempt to transfer an ERC-721 token.
-      IERC721(_token).safeTransferFrom(_from, _to, _param);
-    } else {
-      // Attempt to transfer an ERC-20 token.
-      require(IERC20(_token).transferFrom(_from, _to, _param));
-    }
+    IAsset asset = registry.getAsset(_kind);
+    (bool success, bytes memory data) = address(asset).delegatecall(abi.encodeWithSignature("transferTokens(address,address,uint256,address)", _from, _to, _param, _token));
```
[the remark this answers is not in the document]
Comment (poiresel) [reply protocols/swap/contracts/Swap.sol]: abi.encodeWithSignature(string memory signature, ...) returns (bytes memory): Equivalent to abi.encodeWithSelector(bytes4(keccak256(bytes(signature))), ...), I think they're equivalent: https://solidity.readthedocs.io/en/v0.5.12/units-and-global-variables.html#special-variables-and-functions

What was your suggestion for removing IAsset?

File: protocols/swap/test/Swap.js [hunk 14]
```diff
@@ -81,9 +89,22 @@ contract('Swap', async accounts => {
...
+      const kittyCore = await PartialKittyCoreAsset.new()
+      const erc20Asset = await ERC20Asset.new()
+      const erc721Asset = await ERC721Asset.new()
+      const usdtaAsset = await USDTAsset.new()
+
+      // add all 4 of these contracts into the TokenRegistry
+      tokenRegistry.addToRegistry('0x9a20483d', kittyCore.address)
+      tokenRegistry.addToRegistry('0x277f8169', erc20Asset.address)
+      tokenRegistry.addToRegistry('0x80ac58cd', erc721Asset.address)
+      tokenRegistry.addToRegistry('0xffffffff', usdtaAsset.address)
```
[the remark this answers is not in the document]
Comment (poiresel) [reply protocols/swap/test/Swap.js]: so I didn't generator these selectors, i took them directly from other smart contracts that show this, i'll try to compute it though. Also the last selector, '0xffffffff', is still an open question, still asking for feedback/suggestions.

File: protocols/swap/contracts/Swap.sol (hunk 13 above)
[the remark this answers is not in the document]
Comment (ejwessel) [reply protocols/swap/contracts/Swap.sol]: My suggestion was to keep IAsset and use the selector directly from it like such:
```
(bool success, bytes memory data) = address(asset).delegatecall(IAsset.transferTokens.selector, _from, _to, _param, _token));
```

[the remark this answers is not in the document]
Comment (poiresel) [reply protocols/swap/contracts/Swap.sol]: per your suggestion, ``` Swap.sol:353:7: TypeError: Member "transferTokens" not found or not visible after argument-dependent lookup in type(contract ITransferHandler).
      ITransferHandler.transferTokens.selector,``` is what i'm getting.

[the remark this answers is not in the document]
Comment (poiresel) [reply protocols/swap/contracts/Swap.sol]: ok updated to use it.

File: protocols/swap/test/Swap-unit.js [hunk 15]
```diff
@@ -41,8 +43,13 @@ contract('Swap Unit Tests', async accounts => {
...
   })
 
+  async function setupMockTokenRegistry() {
+    mockTokenRegistry = await MockContract.new()
+  }
+
   before('deploy Swap', async () => {
-    swap = await Swap.new()
+    await setupMockTokenRegistry()
+    swap = await Swap.new(mockTokenRegistry.address)
```
[the remark this answers is not in the document]
Comment (poiresel) [reply protocols/swap/test/Swap-unit.js]: remove it, and no.

**Comment by dmosites** [comment]
Closing until further discussion on the feature.

**Pull request closed without merging**